src/Reflex/Patch/Map.hs

```diff
@@ -22,12 +25,34 @@ newtype PatchMap k v = PatchMap { unPatchMap :: Map k (Maybe v) }
 instance Ord k => Patch (PatchMap k v) where
   type PatchTarget (PatchMap k v) = Map k v
   {-# INLINABLE apply #-}
-  apply (PatchMap p) old = Just $! insertions `Map.union` (old `Map.difference` deletions) --TODO: return Nothing sometimes --Note: the strict application here is critical to ensuring that incremental merges don't hold onto all their prerequisite events forever; can we make this more robust?
-    where insertions = Map.mapMaybeWithKey (const id) p
-          deletions = Map.mapMaybeWithKey (const nothingToJust) p
-          nothingToJust = \case
-            Nothing -> Just ()
-            Just _ -> Nothing
+  apply (PatchMap p) old
+    = changedToMaybe $
+        mergeA
+          (traverseMaybeMissing $ \_k mv ->
+             case mv of
+               Nothing -> Unchanged Nothing
+               Just _ -> Changed mv)
+          preserveMissing
+          -- We could try to detect an update here that does nothing, but that
+          -- will be quite unreliable for a map of Events or similar; it may
+          -- not be worth the trouble.
+          (zipWithMaybeAMatched (\_k mv v -> Changed $! mv <|> Just v)) p old
+
+changedToMaybe :: Changed a -> Maybe a
+changedToMaybe (Unchanged _) = Nothing
+changedToMaybe (Changed a) = Just a
+
+data Changed a
+  = Unchanged a
+  | Changed a
+  deriving (Functor)
```

Comment:
Should I change this to `data Changed a = Changed !Bool a` or similar? I'll have to look at Core to see if GHC will unbox it nicely in that case. Might be an improvement.